java/patterns-use-cases/src/main/java/my/example/cron/CronJob.java

```diff
@@ -0,0 +1,144 @@
+      new CronParser(CronDefinitionBuilder.instanceDefinitionFor(CronType.UNIX));
+
+  @Handler
+  public CronJobSpec initiateJob(ObjectContext ctx, CronRequest req) {
+    // Check if the job already exists
+    var job = ctx.get(JOB);
+    if (job.isPresent()) {
+      throw new TerminalException(
+          "Job already exists. Use a different ID or cancel the existing job first.");
+    }
+
+    // starting the cron job the first time
+    return scheduleNextExecution(ctx, req);
+  }
+
+  @Handler
+  public void execute(ObjectContext ctx) {
+    var job =
+        ctx.get(JOB).orElseThrow(() -> new TerminalException("No cron job information found."));
+
+    // Execute the task
+    ctx.send(createRequest(job.req));
+
+    scheduleNextExecution(ctx, job.req);
+  }
+
+  @Handler
+  public void cancel(ObjectContext ctx) {
+    // Cancel the cron job by canceling the next execution
+    var job = ctx.get(JOB);
+    if (job.isPresent()) {
+      ctx.invocationHandle(job.get().nextExecutionId).cancel();
+    }
+
+    // Clear the job state
+    ctx.clearAll();
+  }
+
+  @Shared
+  public Optional<CronJobSpec> getInfo(SharedObjectContext ctx) {
+    return ctx.get(JOB);
+  }
+
+  private CronJobSpec scheduleNextExecution(ObjectContext ctx, CronRequest req) {
+    // Parse the cron expression to determine the next execution time
+    // Persist current date in Restate for deterministic replay
+    var now = ctx.run(ZonedDateTime.class, ZonedDateTime::now);
+
+    ExecutionTime executionTime;
+    try {
+      executionTime = ExecutionTime.forCron(cronParser.parse(req.expr));
+    } catch (IllegalArgumentException e) {
+      throw new TerminalException("Invalid cron expression " + req.expr + ": " + e.getMessage());
+    }
+
+    // Get delay and next execution time
+    var delay =
+        executionTime
+            .timeToNextExecution(now)
+            .orElseThrow(() -> new TerminalException("No next cron execution time found."));
+    var nextExecutionTime =
+        executionTime
+            .nextExecution(now)
+            .orElseThrow(() -> new TerminalException("No next cron execution time found."));
+
+    // Schedule the job to run at the next execution time
+    var handle = CronJobClient.fromContext(ctx, ctx.key()).send().execute(delay);
+
+    // Store the job spec in the context
+    var job = new CronJobSpec(req, nextExecutionTime.toString(), handle.invocationId());
+    ctx.set(JOB, job);
+    return job;
+  }
+
+  private static Request<byte[], byte[]> createRequest(CronJob.CronRequest req) {
+    ObjectMapper mapper = new ObjectMapper();
+    byte[] payloadBytes =
+        req.parameter
+            .map(
+                param -> {
+                  try {
+                    return mapper.writeValueAsString(param).getBytes();
+                  } catch (JsonProcessingException e) {
+                    throw new TerminalException(e.getMessage());
+                  }
+                })
+            .orElse(new byte[0]);
```

Comment:
Just use serdes here:

If no parameter, then use `Request.of(target, Void.class, Void.class, null)`. If you gotta send the string, use `Request.of(target, String.class, Void.class, parameter)`.